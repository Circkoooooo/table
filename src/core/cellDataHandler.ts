import deepClone from "../tools/deepClone"
import { getColumnLabel, getRowLabel } from "./ruler"

interface FullTableEmptyDataParams {
	rowNum: number
	columnNum: number
}

export type CellDataElement = string | number | null

export type CellData = CellDataElement[][]

export interface WithRulerCellData {
	data: CellData
	info: {
		rowLength: number
		columnLength: number
	}
}
/**
 * 根据row column数目返回一个矩阵 r * c
 * @param params
 */
export const createEmptyCellData = (params: FullTableEmptyDataParams): CellData => {
	const { rowNum, columnNum } = params

	return Array.from(Array(rowNum), () => new Array(columnNum).fill(null))
}

/**
 * 在空二维数组的基础上添加行标和列标（深克隆）
 * @param emptyRulerCellData
 */
export const createRulerCellData = (emptyRulerCellData: CellData): WithRulerCellData => {
	const tempEmptyRulerCellData = deepClone(emptyRulerCellData)

	const result = {
		data: tempEmptyRulerCellData,
		info: {
			rowLength: tempEmptyRulerCellData.length,
			columnLength: (tempEmptyRulerCellData[0] && tempEmptyRulerCellData[0].length) || 0,
		},
	}

	if (tempEmptyRulerCellData.length === 0) return result

	const rowLength = tempEmptyRulerCellData.length
	const columnLength = tempEmptyRulerCellData[0].length

	const columnLabels = getColumnLabel(columnLength, 65, 90)
	const rowLabels = getRowLabel(rowLength, 0)

	const resolvedColumnCellData = [...tempEmptyRulerCellData].map((item, i) => {
		item.unshift(rowLabels[i])
		return item
	})

	//inject one row into
	const rulerCellData = [[null, ...columnLabels], ...resolvedColumnCellData]

	return {
		data: rulerCellData,
		info: {
			rowLength: rulerCellData.length,
			columnLength: rulerCellData[0].length,
		},
	}
}
