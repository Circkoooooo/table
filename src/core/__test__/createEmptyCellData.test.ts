import { createEmptyCellData } from "../cellDataHandler"

test("m * n matrix", () => {
	const result = createEmptyCellData({
		rowNum: 2,
		columnNum: 3,
	})

	expect(result).toEqual([
		[null, null, null],
		[null, null, null],
	])

	expect(result).toMatchSnapshot()
})

export {}
