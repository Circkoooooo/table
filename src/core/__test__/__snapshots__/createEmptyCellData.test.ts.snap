// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`m * n matrix 1`] = `
Array [
  Array [
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
  ],
]
`;
